File: ulohy/26-poklad-v-rece/reseni/main.cs
using System;

internal class Program
{
    static void Main(string[] args)
    {
        int n = ReadInt();
        Console.ReadLine();

        string[] cityNames = new string[n];
        int[] cityLocations = new int[n];

        for (int i = 0; i < n; i++)
        {
            string name = Console.ReadLine();
            int distance = ReadInt();

            cityNames[i] = name;
            cityLocations[i] = distance;
        }

        Console.ReadLine();

        int chestLocation = ReadInt();

        SortByDistance(cityNames, cityLocations, chestLocation);

        for (int i = 0; i < n; i++)
        {
            Console.WriteLine(cityNames[i]);
        }
    }

    static void SortByDistance(string[] cityNames, int[] cityLocations, int chestLocation)
    {
        bool swapped = true;

        while (swapped)
        {
            swapped = false;

            for (int i = 0; i < cityLocations.Length - 1; i++)
            {
                int distance1 = Math.Abs(chestLocation - cityLocations[i]);
                int distance2 = Math.Abs(chestLocation - cityLocations[i + 1]);

                if (distance1 < distance2 || (distance1 == distance2 && cityLocations[i] < cityLocations[i + 1]))
                {
                    continue;
                }

                int tempLocation = cityLocations[i];
                cityLocations[i] = cityLocations[i + 1];
                cityLocations[i + 1] = tempLocation;

                string tempName = cityNames[i];
                cityNames[i] = cityNames[i + 1];
                cityNames[i + 1] = tempName;

                swapped = true;
            }
        }
    }

    static int ReadInt()
    {
        return int.Parse(Console.ReadLine());
    }
}

File: ulohy/26-poklad-v-rece/reseni/main.py
def read_int():
    return int(input())


def sort_by_distance(city_names, city_locations, chest_location):
    swapped = True

    while swapped:
        swapped = False

        for i in range(len(city_locations) - 1):
            distance1 = abs(chest_location - city_locations[i])
            distance2 = abs(chest_location - city_locations[i + 1])

            if distance1 < distance2 or (distance1 == distance2 and city_locations[i] < city_locations[i + 1]):
                continue

            temp = city_locations[i]
            city_locations[i] = city_locations[i + 1]
            city_locations[i + 1] = temp

            temp = city_names[i]
            city_names[i] = city_names[i + 1]
            city_names[i + 1] = temp

            swapped = True


if __name__ == "__main__":
    n = read_int()
    input()

    city_names = []
    city_locations = []

    for _ in range(n):
        city_names.append(input())
        city_locations.append(read_int())

    input()

    chest_location = read_int()

    sort_by_distance(city_names, city_locations, chest_location)

    for city_name in city_names:
        print(city_name)

